crates/red_knot_python_semantic/src/types.rs

```diff
@@ -901,13 +906,16 @@ impl<'db> Type<'db> {
             // Similarly `type[enum.Enum]`  is a subtype of `enum.EnumMeta` because `enum.Enum`
             // is an instance of `enum.EnumMeta`.
             (
-                Type::SubclassOf(SubclassOfType {
-                    base: ClassBase::Class(self_class),
-                }),
+                Type::SubclassOf(subclass_of_ty),
                 Type::Instance(InstanceType {
                     class: target_class,
                 }),
-            ) => self_class.is_instance_of(db, target_class),
+            ) => match subclass_of_ty.subclass_of() {
+                ClassBase::Any | ClassBase::Todo(_) | ClassBase::Unknown => {
+                    unreachable!("Non-fully-static types do not participate in subtyping")
+                }
+                ClassBase::Class(subclass_class) => subclass_class.is_instance_of(db, target_class),
+            },
```

Comment:
Why is this change necessary / an improvement?

Reply:
The existing branch no longer works with this PR because the inner data wrapped by `SubclassOfType` is now private, so we can't simply unpack it as part of the `match` branch. The inner wrapped data must be private if we want to enforce that the eagerly normalising constructor is always used.

But this brand doesn't have to use a nested `match`; it could also be something like `subclass_of_ty.subclass_of().into_class().is_some_and(...)`